src/Microsoft.AspNetCore.Sockets.Client.Http/HttpConnection.cs

```diff
@@ -154,202 +110,179 @@ public HttpConnection(Uri url, ITransportFactory transportFactory, ILoggerFactor
-        {
-            var negotiationResponse = await Negotiate(Url, _httpClient, _logger);
-            _connectionId = negotiationResponse.ConnectionId;
-            _logScope.ConnectionId = _connectionId;
-            return negotiationResponse;
+                await SelectAndStartTransport(transferFormat);
+
+                _started = true;
+                Log.Started(_logger);
+            }
+            finally
+            {
+                _startLock.Release();
+            }
         }
 
-        private async Task StartAsyncInternal(TransferFormat transferFormat)
+        public async Task DisposeAsync() => await DisposeAsyncCore().ForceAsync();
+
+        private async Task DisposeAsyncCore(Exception exception = null)
         {
-            Log.HttpConnectionStarting(_logger);
+            if (_disposed)
+            {
+                return;
+            }
 
+            await _startLock.WaitAsync();
             try
             {
-                var connectUrl = Url;
-                if (_requestedTransportType == TransportType.WebSockets)
+                if (!_disposed && _started)
                 {
-                    // if we're running on Windows 7 this could throw because the OS does not support web sockets
-                    Log.StartingTransport(_logger, _requestedTransportType, connectUrl);
-                    await StartTransport(connectUrl, _requestedTransportType, transferFormat);
-                }
-                else
-                {
-                    var negotiationResponse = await GetNegotiationResponse();
-
-                    // Connection is being disposed while start was in progress
-                    if (_connectionState == ConnectionState.Disposed)
-                    {
-                        Log.HttpConnectionClosed(_logger);
-                        return;
-                    }
+                    Log.DisposingHttpConnection(_logger);
 
-                    // This should only need to happen once
-                    connectUrl = CreateConnectUrl(Url, negotiationResponse.ConnectionId);
+                    // Complete our ends of the pipes.
+                    _transportPipe.Input.Complete(exception);
+                    _transportPipe.Output.Complete(exception);
 
-                    // We're going to search for the transfer format as a string because we don't want to parse
-                    // all the transfer formats in the negotiation response, and we want to allow transfer formats
-                    // we don't understand in the negotiate response.
-                    var transferFormatString = transferFormat.ToString();
-
-                    foreach (var transport in negotiationResponse.AvailableTransports)
+                    // Stop the transport, but we don't care if it throws.
+                    // The transport should also have completed the pipe with this exception.
+                    // REVIEW: We could make not throwing part of the contract here.
```

Comment:
Unless something goes very wrong, throwing in dispose/close feels bad to me.

Reply:
Just to clarify, are you talking about making not throwing part of the `ITransport.StopAsync` contract? That's the exception we're suppressing here. It should be fine but I'll have to review the transport behavior.